### ISSUE TITLE
Dependency gmsh failing on ARM

### ISSUE THREAD

**Issue by Botspot** [l-koehler/FF-converter#11]
Use-case:
STEP files are one type of 3d file and STL files are another. My 3d printer slicer software can only use STL files, but I download STEP files from time to time.
Right now I have to use this website to do it: https://fabconvert.com/convert/step/to/stl

Could you see if this conversion could be done offline remotely easily? If so, this might be a cool feature to add to FF-converter.

**Comment by l-koehler** [comment]
Yes, can be done! I didn't see your issue for a while, but expect a update adding it today/tomorrow.

**Comment by l-koehler** [comment]
so the git version has that conversion added now, but needs a patch to a library.
i won't make a release with trimesh until i got that patch upstream, because requiring users to merge a patch file seems like a bad idea.

**Comment by Botspot** [comment]
> so the git version has that conversion added now, but needs a patch to a library. i won't make a release with trimesh until i got that patch upstream, because requiring users to merge a patch file seems like a bad idea.

Not sure if this is like a system library or something else, but I know that some programs ship with custom libraries for internal usage. That may be an option for your situation, or it may not be.

**Comment by l-koehler** [comment]
its a python library that can handle 3d files, i don't really want to ship a custom library to change two lines.
I already made a pr with which the conversion works for me: https://github.com/mikedh/trimesh/pull/2204

**Comment by l-koehler** [comment]
Added it now, the conversion depends on the python packages `trimesh` and `gmsh`.

**Comment by Botspot** [comment]
> Added it now, the conversion depends on the python packages `trimesh` and `gmsh`.

Please clarify what the effects of this are. Should the pi-apps script for FF-Converter add those 2 packages?

**Comment by Botspot** [comment]
I'm not sure how to correctly install trimesh and gmsh for FFconverter. Could you help?

While I know very little about python, it seems that packages can specify other python packages as dependencies. It seems that ffconverter did not add those as dependencies, so I am trying to install them myself.

(This is after using Pi-Apps to install ff-converter. Pi-Apps uses pipx to install ffconverter.)
```
pi@raspberrypi:~ $ sudo PIPX_HOME=/usr/local/pipx PIPX_BIN_DIR=/usr/local/bin pipx inject ffconverter trimesh gmsh
  injected package trimesh into venv ffconverter
done! ✨ 🌟 ✨
Fatal error from pip prevented installation. Full pip output in file:
    /usr/local/pipx/logs/cmd_2024-04-06_21.53.13_pip_errors.log

Some possibly relevant errors from pip install:
    ERROR: Could not find a version that satisfies the requirement gmsh (from versions: none)
    ERROR: No matching distribution found for gmsh
Error installing gmsh.
```
Any ideas?

**Comment by l-koehler** [comment]
Yes, I didn't add these as dependencies as they aren't strictly required, only enabling a additional conversion.
But that makes packaging with pipx difficult, so trimesh/gmsh are now dependencies in the PyPi release.
pipx should now automatically handle installing trimesh/gmsh (and pyqt5, i forgot adding that to setup.py)

**Comment by l-koehler** [comment]
With pyqt5 added as a dependency, you should be able to change the install script to not install the system package anymore:
```bash
install_packages python3 python3-distutils ffmpeg imagemagick unoconv tar binutils squashfs-tools pandoc || exit 1
pipx_install --system-site-packages "ffconverter==$version" || exit 1
```
pyqt5, trimesh and gmsh should be installed by pipx.
I don't know a lot about pipx, but maybe this also allows you to remove `--system-site-packages`.

**Comment by Botspot** [comment]
> Yes, I didn't add these as dependencies as they aren't strictly required, only enabling a additional conversion. But that makes packaging with pipx difficult, so trimesh/gmsh are now dependencies in the PyPi release. pipx should now automatically handle installing trimesh/gmsh (and pyqt5, i forgot adding that to setup.py)

Well that did not fix the issue of gmsh being missing.
You have effectively broken the app, and our automated update-tester has held your version at 2.4.1 until it is fixed.
See the output here: https://github.com/Botspot/pi-apps/actions/runs/8588579197/job/23533661651#step:7:1314
Looks like it is failing to install on bullseye armhf, bullseye arm64, bookworm armhf, bookworm arm64.

pipx is trying to pull in gmsh, but cannot find it. Where are you getting gmsh from?

**Comment by l-koehler** [comment]
gmsh is available on PyPi, but does not appear to be available for ARM.
That's my bad for not checking, sorry.
I will try to compile it from source on a ARM device and send you a tested install script.

**Comment by l-koehler** [comment]
c7ace8f (v2.4.3) should fix the failing builds by removing trimesh/gmsh from the dependencies and only offering a optional dependency on x86_64.
There is no simple way for me to provide a version of GMSH compiled for ARM, so users needing this library will need to compile it.
I added instructions for compiling GMSH on ARM to the README.md. Note that GMSH is a fairly big dependency and should not be re-compiled on every update.
Consider checking the exit code of `python3 -c "import sys;sys.path.append('/usr/local/lib');import gmsh"`, this should be `0` only if GMSH is already installed.

pi-apps will need to do the following:
* install ff-converter with access to system packages (just what you did in 2.3.x)
* install trimesh with pip
* compile gmsh if it isn't yet installed

Also consider letting the user opt into getting gmsh installed, increasing the install time from seconds to hours for something most people won't use seems annoying (trimesh remains useful without gmsh and is available over pip on arm, so that can still be a default).

**Comment by l-koehler** [comment]
I would close this, as there is no alternative to GMSH that offers prebuilt ARM packages.

The description for ff-converter could link to [this part of the readme](https://github.com/l-koehler/ff-converter?tab=readme-ov-file#troubleshooting-arm-cpus), it explains how to install GMSH if a user needs STEP file support.

I also noticed the install script for ff-converter does not install the python package "trimesh", adding that would add support for some 3D files even without GMSH.
